PSReadLine/PlatformWindows.cs

```diff
@@ -614,4 +611,350 @@ public override void BlankRestOfLine()
+
+            [PreserveSig]
+            int AddTab( IntPtr hwnd );
+
+            [PreserveSig]
+            int DeleteTab( IntPtr hwnd );
+
+            [PreserveSig]
+            int ActivateTab( IntPtr hwnd );
+
+            [PreserveSig]
+            int SetActiveAlt( IntPtr hwnd );
+
+            // ITaskbarList2
+            [PreserveSig]
+            int MarkFullscreenWindow( IntPtr hwnd, [MarshalAs(UnmanagedType.Bool)] bool fFullscreen );
+
+            // ITaskbarList3
+            [PreserveSig]
+            int SetProgressValue( IntPtr hwnd, UInt64 ullCompleted, UInt64 ullTotal );
+
+            [PreserveSig]
+            int SetProgressState( IntPtr hwnd, TaskbarStates state );
+
+            // N.B. for copy/pasters: we've left out the rest of the ITaskbarList3 methods...
+        }
+
+        [ComImport()]
+        [Guid("56fdf344-fd6d-11d0-958a-006097c9a090")]
+        [ClassInterface(ClassInterfaceType.None)]
+        private class TaskbarInstance
+        {
+        }
+
+        private static Lazy<ITaskbarList3> _taskbarInstance = new Lazy<ITaskbarList3>(() => (ITaskbarList3) new TaskbarInstance());
+
+        public static int SetProgressState(IntPtr windowHandle, TaskbarStates taskbarState)
+        {
+            return _taskbarInstance.Value.SetProgressState(windowHandle, taskbarState);
+        }
+
+        public static int SetProgressValue(IntPtr windowHandle, int progressValue, int progressMax)
+        {
+            return _taskbarInstance.Value.SetProgressValue(windowHandle, (ulong) progressValue, (ulong) progressMax);
+        }
+    }
+
+    // Calculates what processes need to be terminated (populated into procsToTerminate),
+    // and returns the count. A "straggler" is a console-attached process (so GUI
+    // processes don't count) that is not in the _allowedPids list.
+    private static int GatherStragglers(List<Process> procsToTerminate)
+    {
+        procsToTerminate.Clear();
+
+        // These are the processes currently attached to this console. Note that GUI
+        // processes will not be attached to the console.
+        uint[] currentPids = GetConsoleProcessList();
+
+        foreach (var pid in currentPids)
+        {
+            if (!_allowedPids.Contains(pid))
+            {
+                Process proc = null;
+                try
+                {
+                    proc = Process.GetProcessById((int) pid);
+                }
+                catch( ArgumentException )
+                {
+                    // Ignore it: process could be gone, or something else that we
+                    // likely can't do anything about it.
+                }
+
+                if (proc != null)
+                {
+                    procsToTerminate.Add(proc);
+                }
+            }
+        }
+        return procsToTerminate.Count;
+    }
+
+    [DllImport("kernel32.dll")]
+    internal static extern ulong GetTickCount64();
+
+    private const int DefaultGraceMillis = 1000;
```

Comment:
>DefaultGraceMillis = 1000; [](http://example.com/codeflow?start=22&length=26)

We *could* make this configurable... but I don't think it's necessary; and the downside is that it's more cognitive load for the user ("what is this?")